static/app/views/dashboardsV2/detail.tsx

```diff
@@ -724,3 +697,49 @@ const StyledPageContent = styled(PageContent)`
 `;
 
 export default withApi(withOrganization(DashboardDetail));
+
+/**
+ * Reads the layout from an array of widgets.
+ */
+function getDashboardLayout(widgets: Widget[]): RGLLayout[] {
+  return widgets
+    .filter(({layout}) => !!layout)
+    .map(({layout, ...widget}) => ({
+      ...(layout as RGLLayout),
+      i: constructGridItemKey(widget),
+    }));
+}
+
+// Keys for grid layout values we track in the server
+const STORE_KEYS = ['x', 'y', 'w', 'h', 'minW', 'maxW', 'minH', 'maxH'];
+
+/**
+ * Creates a new DashboardDetails object with the layouts associated with
+ * widgets for outgoing requests.
+ */
+function constructDashboardWidgetsWithLayout(
+  dashboard: DashboardDetails,
+  layout: RGLLayout[]
+): DashboardDetails {
+  return {
+    ...dashboard,
+    widgets: dashboard.widgets.map(widget => {
+      const matchingLayout = layout.find(({i}) => i === constructGridItemKey(widget));
+      return {...widget, layout: pick(matchingLayout, STORE_KEYS)};
+    }),
+  };
+}
+
+function isLayoutEqual(prevLayouts: RGLLayout[], newLayouts: RGLLayout[]): boolean {
+  // Compares only defined keys we care about storing
+  const normalizeLayout = layout => {
+    const definedKeys = Object.keys(layout).filter(
+      key => STORE_KEYS.includes(key) && layout[key] !== undefined
+    );
+    return pick(layout, definedKeys);
```

Comment:
Do we need to filter `definedKeys` first?

```
> var object = { 'a': 1, 'b': '2', 'c': 3 };
undefined
> lodash.pick(object, ['a', 'c', 'f'])
{ a: 1, c: 3 }
```

Reply:
I don't think so. If they're unset then that's fine, we only care about comparing the values that are set and that we keep track of in the backend